src/utils/bounded_priority_queue.rs

```diff
@@ -0,0 +1,57 @@
+use crate::serializable_traits::Data;
+use serde_derive::{Deserialize, Serialize};
+use std::collections::BinaryHeap;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct BoundedPriorityQueue<T: Ord> {
+    max_size: usize,
+    underlying: BinaryHeap<T>,
+}
+
+impl<T: Data + Ord> Into<Vec<T>> for BoundedPriorityQueue<T> {
+    fn into(self) -> Vec<T> {
+        let mut col: Vec<_> = self
+            .underlying
+            .into_iter_sorted()
+            .collect();
+        col.reverse();
+        col
+    }
+}
+
+impl<T: Data + Ord> BoundedPriorityQueue<T> {
+    pub fn new(max_size: usize) -> BoundedPriorityQueue<T> {
+        BoundedPriorityQueue {
+            max_size: max_size,
+            underlying: BinaryHeap::with_capacity(max_size),
+        }
+    }
+
+    pub fn merge(&mut self, other: BoundedPriorityQueue<T>) -> &Self {
```

Comment:
I think conceptually here is better to take self by move since:
```
pub fn merge(mut self, other: BoundedPriorityQueue<T>) -> Self {
       other
            .underlying
            .into_iter()
            .for_each(|elem| self.append(elem));
        self
}
```

then on `take_ordered` you can do:
```
            let queue = self
                .map_partitions(first_k_func)
                .reduce(Fn!(
                    move |mut queue1: BoundedPriorityQueue<Self::Item>,
                          queue2: BoundedPriorityQueue<Self::Item>|
                          -> BoundedPriorityQueue<Self::Item> {
                        queue1.merge(queue2)
                    }
                ))?
                .unwrap() as BoundedPriorityQueue<Self::Item>;
```

Or at the very least you should be returning `&mut self`, this is not a problem in this particular context, but it could turn annoying in other circumstances (returning a & instead of a &mut after taking &mut self).

Reply:
Wow the `move` and `ownership` is such subtle concepts...and it starts to make sense to me now. Thanks, will try.